cmd/query/app/server.go

```diff
@@ -114,16 +125,81 @@ func createHTTPServer(querySvc *querysvc.QueryService, queryOpts *QueryOptions,
 	recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
 	return &http.Server{
 		Handler: recoveryHandler(handler),
+	}, nil
+}
+
+func getTLSListener(listener net.Listener, tlsOptions tlscfg.Options) (net.Listener, error) { // takes otherCA  so that the certPool will have the CA of both GRPC and HTTP clients.
+	tlsCfg, err := tlsOptions.Config()
+	if err != nil {
+		return nil, err
+	}
+	tlsCfg.Rand = rand.Reader
+	tlsListener := tls.NewListener(listener, tlsCfg)
+	return tlsListener, err
+
+}
+
+func (s *Server) getCmux() (cmux.CMux, net.Listener, net.Listener, error) {
+	var httpListener net.Listener
+	var grpcListener net.Listener
+	var cmux1 cmux.CMux
+	conn, err := net.Listen("tcp", s.queryOptions.HostPort)
+	s.conn = conn
+
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if s.queryOptions.HTTPTLSEnabled != s.queryOptions.GRPCTLSEnabled { // exactly one of HTTP or GRPC has TLS enabled
+		cmux1 = cmux.New(conn)
+
+		if !s.queryOptions.HTTPTLSEnabled { // HTTPTLS disabled GRPCTLS enabled => match http request
+			httpListener = cmux1.Match(cmux.HTTP1Fast())
```

Comment:
If I'm reading this right, using a HTTP2 client will then get `cmux` to serve the request using TLS.

Honestly, I'm starting to think that `cmux` isn't worth the complexity it brings. Debugging an issue around this code is probably going to be a nightmare...